src/compiler/checker.ts

```diff
@@ -4929,6 +4998,12 @@ module ts {
                 if (source === anyFunctionType) {
                     return;
                 }
+                if (target.flags & TypeFlags.TypeGuard) {
+                    target = (<TypeGuardType>target).type;
+                }
+                if (source.flags & TypeFlags.TypeGuard) {
+                    source = (<TypeGuardType>source).type;
+                }

```

Comment:
This does not seem correct.
```ts
declare function isFoo(x): x is Foo;
declare function generic<T>(func: (p) => T): T;
generic(isFoo);
```
Here you will infer Foo for T, but really I think you should infer boolean.